riscv-executor/src/profiler.rs

```diff
@@ -0,0 +1,346 @@
+    /// This is handled as a "call" into a function.
+    pub fn jump_and_link(&mut self, curr_pc: usize, target_pc: usize, return_pc: usize) {
+        if !self.options.is_enabled() {
+            return;
+        }
+        if let Some(mut target) = self.location_at(target_pc) {
+            if let Some(curr_function) = self.curr_function() {
+                let Loc {
+                    file: curr_file,
+                    line: curr_line,
+                    ..
+                } = self.location_at(curr_pc).unwrap();
+                // ecall handler code have a ".debug loc", so we keep current file/line
+                if target.function == "__ecall_handler" {
+                    target.file = curr_file;
+                    target.line = curr_line;
+                }
+                let call = Call {
+                    from: Loc {
+                        function: curr_function,
+                        file: curr_file,
+                        line: curr_line,
+                    },
+                    target,
+                };
+                // increase call count
+                self.call_stats.entry(call.clone()).or_default().0 += 1;
+                self.call_stack.push((call, 0));
+                self.return_pc_stack.push(return_pc);
+            } else {
+                // we start profiling on the initial call to "__runtime_start"
+                if target.function == "__runtime_start" {
+                    let call = Call {
+                        from: Loc {
+                            function: "",
+                            file: 0,
+                            line: 0,
+                        },
+                        target,
+                    };
+                    // increase call count
+                    self.call_stats.entry(call.clone()).or_default().0 += 1;
+                    self.call_stack.push((call, 0));
+                    self.return_pc_stack.push(return_pc);
+                }
+            }
+        } else {
+            assert!(!self.running());
+        }
+    }
+
+    /// Should be called for jumps that don't save the returning address.
+    /// This is handled as one of 3 cases:
+    /// - "return" from function: target_pc equal to last `jump_and_link` saved pc
+    /// - "tail call": next_function != current_function
+    /// - control flow: next_function == current_function
+    pub fn jump(&mut self, target_pc: usize) {
+        if !self.options.is_enabled() || !self.running() {
+            return;
+        }
+
+        if self
+            .return_pc_stack
+            .last()
+            .is_some_and(|saved_pc| *saved_pc == target_pc)
+        {
+            // "return" from current function
+            let (done_call, cost) = self.call_stack.pop().unwrap();
+            self.return_pc_stack.pop();
+            // add to calls cummulative cost and to parent call
```

Comment:
```suggestion
            // add to calls cumulative cost and to parent call
```

Reply:
wdym by `to parent call`?

Reply:
we update two costs here (e.g., lets say this is `bar` calling `foo` from line 20):
- cumulative cost of this specific call (`foo -> bar` at line 20)
- running cost of caller (`foo`)